### PR TITLE
[Executor][BatchEngine] Support init in script executor

### PR DESCRIPTION
**Pull request by D-W-**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

Support the following usage:

### CLI

```bash
pf run create --flow "./flow.dag.yaml" --inits chat_model_deployment_name="gpt-35-turbo" --data ./data.jsonl

```

## serve

```bash
pf flow serve --source "./"  --port 8088 --host localhost --inits chat_model_deployment_name="gpt-35-turbo"
```

for class based flows:

```python
class MyFlow:
    def __init__(self, obj_input: str):
        pass

    def __call__(self, func_input: str) -> dict:
        pass

```

This pull request introduces a new `init` parameter to several methods and classes in the `promptflow` package, which is used to initialize callable classes in Python scripts. The `init` parameter is a dictionary that contains the initialization arguments for the callable class. The changes are spread across different files in the `promptflow` and `tests` directories.

Major changes include:

Initialization of callable classes:

* [`src/promptflow/promptflow/executor/_script_executor.py`](diffhunk://#diff-3294a0ccdf18c47385db91932bd860940cc3df5df10d206138fbcd48075530d5R36-R42): Added the `init` parameter to the `__init__` method and `_initialize_function` method of the `ScriptExecutor` class. If the function to be executed is a callable class, it is initialized with the `init` arguments. [[1]](diffhunk://#diff-3294a0ccdf18c47385db91932bd860940cc3df5df10d206138fbcd48075530d5R36-R42) [[2]](diffhunk://#diff-3294a0ccdf18c47385db91932bd860940cc3df5df10d206138fbcd48075530d5L132-R147)
* [`src/promptflow/promptflow/executor/flow_executor.py`](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69R180): Added the `init` parameter to the `create` method of the `FlowExecutor` class. If the flow is a script flow, the `ScriptExecutor` is created with the `init` arguments. [[1]](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69R180) [[2]](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69R200-R211)

Changes in the proxy factory and executor proxies:

* [`src/promptflow/promptflow/_proxy/_proxy_factory.py`](diffhunk://#diff-936478766b9de290e8791c802585d3932cfed416ae7b50daa76396dacf88cdf5L43-R43): Added the `init` parameter to the `create_executor_proxy` method of the `ProxyFactory` class and passed it to the executor proxy class. [[1]](diffhunk://#diff-936478766b9de290e8791c802585d3932cfed416ae7b50daa76396dacf88cdf5L43-R43) [[2]](diffhunk://#diff-936478766b9de290e8791c802585d3932cfed416ae7b50daa76396dacf88cdf5R52)
* `src/promptflow/promptflow/batch/_base_executor_proxy.py`, `src/promptflow/promptflow/batch/_csharp_executor_proxy.py`, `src/promptflow/promptflow/batch/_python_executor_proxy.py`: Added the `init` parameter to the `create` method of the executor proxy classes. [[1]](diffhunk://#diff-44ac8eeb30a630bd24987e92f8bd5a180d57a3ad067db5d937561f21771e14c5R129) [[2]](diffhunk://#diff-f8219170007d8a89eed104a166ab8c0c9d3af5b5c5e61225659d982f470aeb95R102-R106) [[3]](diffhunk://#diff-8c42522c7186705de9a6651343af6006be02b7342f72467c2785f23f93ea669eR57-R62)

Changes in the batch engine:

* [`src/promptflow/promptflow/batch/_batch_engine.py`](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80R79): Added the `init` parameter to the `__init__` method and `run` method of the `BatchEngine` class, and stored it as an instance variable. The `init` arguments are passed to the executor proxy when it is created. [[1]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80R79) [[2]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80R98-R99) [[3]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80R130) [[4]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80R176)

Changes in the line execution process pool:

* [`src/promptflow/promptflow/executor/_line_execution_process_pool.py`](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1R137-R138): Added the `init` arguments to the initialization arguments of the script executor.

Changes in the tests:

* [`src/promptflow/tests/executor/e2etests/test_eager_flow.py`](diffhunk://#diff-0f88413b3a6799b404c8406e28d7e95a18a99771f78a65cabaaf2ac5e0a6ec8bL49-R83): Added test cases for flows with callable classes and initialization arguments. [[1]](diffhunk://#diff-0f88413b3a6799b404c8406e28d7e95a18a99771f78a65cabaaf2ac5e0a6ec8bL49-R83) [[2]](diffhunk://#diff-0f88413b3a6799b404c8406e28d7e95a18a99771f78a65cabaaf2ac5e0a6ec8bL86-R123) [[3]](diffhunk://#diff-0f88413b3a6799b404c8406e28d7e95a18a99771f78a65cabaaf2ac5e0a6ec8bR157-R191)

Addition of test data:

* [`src/promptflow/tests/test_configs/eager_flows/basic_callable_class/flow.dag.yaml`](diffhunk://#diff-cbdde7417cc8a09c3cc12f4d1406271524ef5efb59d09c81d4dc72540812d6a0R1): Added a new test flow that uses a callable class.
* [`src/promptflow/tests/test_configs/eager_flows/basic_callable_class/inputs.jsonl`](diffhunk://#diff-563fd405c04629603f47fa4f1d0f43afe24cc141182edcc6060a0b8bd5aa02ebR1-R2): Added input data for the new test flow.
* [`src/promptflow/tests/test_configs/eager_flows/basic_callable_class/simple_callable_class.py`](diffhunk://#diff-0cee1cf527fa9d1bfdf418097c8012e3fa1c76b465fc235e92047800d43f77a7R1-R21): Added a new Python script that defines a callable class.
# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_proxy/_proxy_factory.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def register_executor(cls, language: str, executor_proxy_cls: Type[AbstractExecu
         cls.executor_proxy_classes[language] = executor_proxy_cls
 
     def create_executor_proxy(
-        self, flow_file, working_dir, connections, storage, language: str, **kwargs
+        self, flow_file, working_dir, connections, storage, language: str, init: dict, **kwargs
```
Comment (thy09) [inline src/promptflow/promptflow/_proxy/_proxy_factory.py:43]: init_kwargs?

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 2]
```diff
@@ -134,6 +134,8 @@ def __init__(
             "working_dir": flow_executor._working_dir,
             "line_timeout_sec": self._line_timeout_sec,
             "raise_ex": False,
+            # only script executor has init
+            "init": flow_executor._init if hasattr(flow_executor, "_init") else {},
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:138]: Maybe `getattr(flow_executor, "_init_kwargs", {})`?

File: src/promptflow/promptflow/executor/_script_executor.py [hunk 3]
```diff
@@ -129,7 +132,19 @@ def _initialize_function(self):
         module_name, func_name = self._parse_flow_file()
         module = importlib.import_module(module_name)
         func = getattr(module, func_name, None)
-        if func is None or not inspect.isfunction(func):
+        # check if func is a callable class
+        if inspect.isclass(func):
+            if hasattr(func, "__call__"):
+                logger.debug(f"Python class entry '{func_name}' has __call__ method, initializing it with {self._init}")
+                obj = func(**self._init)
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_script_executor.py:139]: We need to raise UserError here if the user function failed.

File: src/promptflow/promptflow/batch/_batch_engine.py [hunk 4]
```diff
@@ -94,6 +95,8 @@ def __init__(
         :type line_timeout: Optional[int]
         :param worker_count: The concurrency limit of batch run
         :type worker_count: Optional[int]
+        :param init: Class init arguments for callable class, only supported for flex flow.
+        :type inits Optional[Dict[str, Any]]
```
Comment (wangchao1230) [inline src/promptflow/promptflow/batch/_batch_engine.py:99]: inits  -> init

Comment (D-W-) [reply src/promptflow/promptflow/batch/_batch_engine.py]: Updated to init_kwargs

**Review by thy09 — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**